File: AlgorithmStudy/src/aug/Solution0805.java
package aug;

import java.util.ArrayList;
import java.util.List;

// https://leetcode.com/explore/challenge/card/august-leetcoding-challenge-2021/613/week-1-august-1st-august-7th/3838/
public class Solution0805 {
    List<List<Integer>> result = new ArrayList<>();

    public List<List<Integer>> pathSum(TreeNode root, int targetSum) {

        subSum(root, targetSum, new ArrayList<>());
        // 경로 전체 탐색
        return result;
        // bfs -> queue
        // dfs -> stack
    }

    private void subSum(TreeNode node, int targetSum, List<Integer> path) {
        if (node == null) {
            return ;
        }

        path.add(node.val);
        // 마지막 노드 인데,
        if (node.left == null && node.right == null && node.val == targetSum)  {
            result.add(new ArrayList<>(path));
        } else {
            subSum(node.left, targetSum - node.val, path);
            subSum(node.right, targetSum - node.val, path);
        }
        //
        path.remove(path.size() - 1);
    }

}

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

File: AlgorithmStudy/src/y2023/aug/Solution0805.kt
package y2023.aug

import model.TreeNode


// https://leetcode.com/problems/unique-binary-search-trees-ii/
class Solution0805 {
    fun generateTrees(n: Int): List<TreeNode?> {
        return dfs(1, n)

    }

    fun dfs(start: Int, end: Int): List<TreeNode?> {
        val result = mutableListOf<TreeNode?>()

        if (start > end) {
            result.add(null)
            return result
        }

        if (start == end) {
            result.add(TreeNode(start))
            return result
        }

        for (i in start .. end) {
            val left = dfs(start, i - 1)
            val right = dfs(i + 1, end)

            for (l in left) {
                for (r in right) {
                    val root = TreeNode(i)
                    root.left = l
                    root.right = r
                    result.add(root)
                }
            }
        }
        return result
    }
}